mythril_core/src/vga_logger.rs

```diff
@@ -0,0 +1,101 @@
+use core::fmt;
+
+const VGA_BASE_ADDR: usize = 0xB8000;
+const VGA_WIDTH: usize = 80;
+const VGA_HEIGHT: usize = 25;
+const VGA_ATTRIB: u16 = 0x0F00; // black background, white text
+
+fn get_vga_addr(x: usize, y: usize) -> usize {
+    VGA_BASE_ADDR + 2 * (y * VGA_WIDTH + x)
+}
+
+unsafe fn scroll_vga() {
+    for y in 1..VGA_HEIGHT {
+        for x in 0..VGA_WIDTH {
+            let write_addr = get_vga_addr(x, y - 1) as *mut u16;
+            let read_addr = get_vga_addr(x, y) as *const u16;
+
+            *write_addr = *read_addr;
+        }
+    }
+    clear_line_vga(VGA_HEIGHT - 1);
+}
+
+unsafe fn clear_line_vga(y: usize) {
+    for x in 0..VGA_WIDTH {
+        let addr = get_vga_addr(x, y) as *mut u16;
+        *addr = VGA_ATTRIB | 0x20;
+    }
+}
+
+pub unsafe fn clear_vga() {
+    for y in 0..VGA_HEIGHT {
+        clear_line_vga(y);
+    }
+}
+
+pub unsafe fn raw_write_vga(s: &str, x: usize, y: usize) -> (usize, usize) {
+    let mut x = x;
+    let mut y = y;
+
+    for byte in s.bytes() {
+        // move cursor on newlines (0x0A) and carriage-returns (0x0D)
```

Comment:
Is this the expected behavior? That newline does a newline and carriage return, while carriage return just goes to the start of the line? Is this from an existing implementation?

Reply:
Since the rust toolchain uses linux-style line returns for `writeln!`, we need to both do a newline and a carriage return so that the lines don't start going off diagonally on the screen. I've added in the check for just carriage returns as well just in case this was built on windows or if a log statement used a carriage return for something like a loading bar.